concurrency chapter

japaric opened the pull request:
This PR adds a chapter that documents the most common (memory) safe uses of `unsafe` code in concurrent bare metal (`no_std`) programs. The patterns described here are used to implement the safe concurrency abstractions provided by the `cortex-m`, `cortex-m-rt` and `cortex-m-rtfm` crates.

[Rendered](https://japaric.github.io/embedonomicon/concurrency.html)

@RalfJung would it be possible to get the UCG WG to (at some point) review / audit these patterns? I'm particularly concerned about LLVM attributes / semantics not matching the intended semantics.

cc @jamesmunns this is relevant to your recent work on shared-rs

japaric committed:
concurrency chapter

rust-highfive commented:
r? @korken89

(rust_highfive has picked a reviewer for you, use r? to override)

therealprof commented on src/concurrency.md:
````diff
@@ -0,0 +1,722 @@
+{{#include ../ci/concurrency/examples/systick.rs}}```
+
+If you are not familiar with embedded / Cortex-M programs the most important
+thing to point note here is that the function marked with the `entry` attribute
````
Should probably either be point out or note.

therealprof commented on src/concurrency.md:
```diff
@@ -0,0 +1,722 @@
+
+As the runtime initialization pattern is used to initialize the "state" of
+interrupt handlers so all the additional constraints required for multi-core
+memory safety of the State pattern are also required here.
```
```suggestion
memory safety of the state pattern are also required here.
```

therealprof commented:
Other than those two nits LGTM

andre-richter commented:
I think the end of the document is missing a subchapter to conclude the multi-core story and therefore the whole document nicely.

First, we tell that Mutex as implemented in the current form is not `Sync`. Then we introduce Atomics as an alternative, and only casually mention they can be used to build spinlocks too.

IMO, we should showcase exactly that after the Atomics chapter. We can probably cite a minimal implementation, as for example it can be found in the [spin crate](https://github.com/mvdnes/spin-rs/blob/master/src/mutex.rs#L131).

Other than that, I like :+1:  :)

korken89 commented:
I'm happy with this, but lets include @therealprof comments. After that I will merge.

korken89 commented:
Something seems amiss with the asm format, can you have a look @japaric ?

RalfJung commented:
> would it be possible to get the UCG WG to (at some point) review / audit these patterns? I'm particularly concerned about LLVM attributes / semantics not matching the intended semantics.

I guess? If you have concrete questions, I suggest to open an issue at https://github.com/rust-rfcs/unsafe-code-guidelines/.  However, in this PR I see 1700 lines of code and I am not sure what to look at.^^

therealprof committed:
Update src/concurrency.md

Co-Authored-By: japaric <jorge@japaric.io>

japaric committed:
fix typo

japaric committed:
preemption matters

also note the bounds and multi-core requirements of the cooperative handlers
pattern

japaric committed:
define MutEx and spinlock

japaric commented:
@therealprof thanks for the review; I have addressed your comments

@andre-richter I have added some notes; let me know if you think something else needs to be clarified

@korken89 if you mean the CI failure; it should have been fixed by #47 

@RalfJung great, I'll open a few issues over there

Disasm commented:
@japaric In **Priorities** section I would suggest using some different interrupts as an example:
* Both `SysTick` and `SVCall` have fixed (and different) priorities.
* They are both exceptions, not interrupts. That's a bit misleading.

jounathaen commented:
I'd like to bump this PR, because the chapter contains a lot of useful information and it would be great to have them in the "documentation"!

jamesmunns commented:
Note: will probably need to be rebased on https://github.com/rust-embedded/embedonomicon/pull/66

jonas-schievink commented on src/concurrency.md:
````diff
@@ -0,0 +1,760 @@
+[this repository]: https://github.com/rust-embedded/embedonomicon
+
+``` rust
+{{#include ../ci/concurrency/examples/systick.rs}}```
````
The backticks on the same line break with the current mdbook version (there's multiple instances of this).

`````suggestion
{{#include ../ci/concurrency/examples/systick.rs}}
```
`````

jonas-schievink commented on src/concurrency.md:
````diff
@@ -0,0 +1,760 @@
+
+``` rust
+#[link_name = "SysTick"] // places this function in the vector table
+fn randomly_generated_identifier() {
````
This is no longer random

jonas-schievink commented on ci/concurrency/examples/init.rs:
```diff
@@ -0,0 +1,65 @@
+// source: examples/init.rs
+
+#![feature(maybe_uninit)]
```
This is stable now

jonas-schievink commented on ci/concurrency/examples/mutex.rs:
```diff
@@ -0,0 +1,54 @@
+
+struct Mutex<T>(UnsafeCell<T>);
+
+// TODO does T require a Sync / Send bound?
+unsafe impl<T> Sync for Mutex<T> {}
```
```suggestion
unsafe impl<T: Send> Sync for Mutex<T> {}
```

jonas-schievink commented on src/concurrency.md:
```diff
@@ -0,0 +1,760 @@
+  contexts* that may preempt each other.
+
+- `Send`: types that can be transferred between *execution contexts* that may
+  preempt each other.
```
It might make sense to update this wrt. the RFC to restrict Send/Sync reasoning to single cores

jonas-schievink commented on src/concurrency.md:
```diff
@@ -0,0 +1,760 @@
+The problem with `Mutex` is not the critical section that uses; it's the fact
+that it can be stored in a `static` variable making accessible to all cores.
+Thus in multi-core context the `Mutex` abstraction should not implement the
+`Sync` trait.
```
Oh yeah this is outdated with the RFC now

jonas-schievink reviewed (commented):
Great writeup!

TDHolmes commented:
Any chance this will be going in at some point? Sounds very useful. I'm coming here from the ["under the hood"](https://rtic.rs/0.5/book/en/internals.html) section of the RTIC documentation
